src/common/quantile.cc

```diff
@@ -89,40 +91,94 @@ std::vector<bst_feature_t> HostSketchContainer::LoadBalance(
   return cols_ptr;
 }
 
-void HostSketchContainer::PushRowPage(SparsePage const &page,
-                                      MetaInfo const &info) {
+namespace {
+// Function to merge hessian and sample weights
+std::vector<float> MergeWeights(MetaInfo const &info,
+                                std::vector<float> const &hessian,
+                                bool use_group, int32_t n_threads) {
+  CHECK_EQ(hessian.size(), info.num_row_);
+  std::vector<float> results(hessian.size());
+  auto const &group_ptr = info.group_ptr_;
+  if (use_group) {
+    auto const &group_weights = info.weights_.HostVector();
+    CHECK_GE(group_ptr.size(), 2);
+    CHECK_EQ(group_ptr.back(), hessian.size());
+    size_t cur_group = 0;
+    for (size_t i = 0; i < hessian.size(); ++i) {
+      results[i] = hessian[i] * group_weights[cur_group];
+      if (i == group_ptr[cur_group + 1]) {
+        cur_group++;
+      }
+    }
```

Comment:
This portion is sequential for now?

Reply:
It's either sequential or binary search.